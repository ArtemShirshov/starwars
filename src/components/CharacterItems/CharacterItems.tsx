import React from 'react';
import { Box, Skeleton, Grid } from '@mui/material';
import Pagination from '@mui/material/Pagination';

import { CharacterItemsViewModel } from './CharacterItems.viewmodel';
import { CharacterItem } from './CharacterItem/CharacterItem';
import { CharacterItemSkeleton } from './CharacterItem/CharacterItemSkeleton';

export const CharacterItems = () => {
  const {
    loading,
    characterItems,
    handleChangePage,
    page,
    onClick,
    characterPages,
  } = CharacterItemsViewModel();

  return (
    <>
      <Grid container spacing={2}>
        {loading
          ? Array.from({
              length: 10,
            }).map((item, index) => (
              <Grid item key={index} md={3} xs={12}>
                <CharacterItemSkeleton />
              </Grid>
            ))
          : characterItems?.map((item: any) => (
              <Grid
                item
                key={item.name}
                md={3}
                xs={12}
                onClick={() => onClick(item.id)}
              >
                <CharacterItem {...item} />
              </Grid>
            ))}
      </Grid>

      <Box
        my={2}
        sx={{
          display: 'flex',
          justifyContent: 'center',
        }}
      >
        {loading && characterPages === 0 ? (
          <Skeleton height={32} width={150} />
        ) : (
          <Pagination
            color="primary"
            count={characterPages}
            onChange={handleChangePage}
            page={page}
            shape="rounded"
          />
        )}
      </Box>
    </>
  );
};
