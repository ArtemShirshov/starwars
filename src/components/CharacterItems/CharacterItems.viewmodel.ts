import { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router';
import { useSearchParams } from 'react-router-dom';

import { peopleService } from 'services/People.service';

import {
  CharacterType,
  getCharacterItems,
  getCharacterPages,
  setCharacterItems,
  setPages,
} from 'store/CharacterItems';

export const CharacterItemsViewModel = () => {
  const [page, setPage] = useState<number | string>(1);
  const [loading, setLoading] = useState(true);

  const [searchParams, setSearchParams] = useSearchParams();
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const characterItems = useSelector(getCharacterItems);
  const characterPages = useSelector(getCharacterPages);

  const fetchCharacter = useCallback(
    ({ page }: { page?: number | string } = {}) => {
      setLoading(true);

      try {
        peopleService
          .getPeople({
            params: {
              page,
            },
          })
          .then(({ data }) => {
            const _data = data.results.map((item: CharacterType) => {
              const match = item.url.match(/\/([^/]+)\/?$/);
              const id = match ? match[1] : null;
              return {
                ...item,
                id,
              };
            });
            dispatch(setCharacterItems(_data));
            dispatch(setPages(Math.ceil(data.count / 10)));
          })
          .finally(() => {
            setLoading(false);
          });
      } catch (error) {
        setLoading(false);
        console.log(error);
      }
    },
    [dispatch]
  );

  const handleChangePage = useCallback(
    (e: any, page: string | number) => {
      setPage(page);
      fetchCharacter({
        page,
      });
      setSearchParams({
        page: String(page),
      });
    },
    [fetchCharacter, setSearchParams]
  );

  const onClick = useCallback(
    (id: string) => {
      navigate(`character/${id}`);
    },
    [navigate]
  );

  useEffect(() => {
    const page = searchParams.get('page');

    fetchCharacter({
      page,
    });

    if (page) {
      setPage(Number(page));
    }

    return () => {
      dispatch(setCharacterItems([]));
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return {
    loading,
    characterItems,
    characterPages,
    handleChangePage,
    page,
    onClick,
  } as const;
};
