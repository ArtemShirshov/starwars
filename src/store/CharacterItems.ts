import type { PayloadAction } from '@reduxjs/toolkit';
import { createSelector, createSlice } from '@reduxjs/toolkit';

export interface CharacterType {
  name: string;
  height: string;
  mass: string;
  hair_color: string;
  skin_color: string;
  eye_color: string;
  birth_year: string;
  gender: string;
  homeworld: string;
  films: string[];
  species: string[];
  vehicles: string[];
  starships: string[];
  created: string;
  edited: string;
  url: string;
  id: string;
}

export interface CharacterItemsStateType {
  items: CharacterType[];
  pages: number;
}

const initialState: CharacterItemsStateType = {
  items: [],
  pages: 0,
};

export const characterItemsSlice = createSlice({
  name: 'characterItems',
  initialState,
  reducers: {
    setCharacterItems: (
      state: CharacterItemsStateType,
      action: PayloadAction<any[]>
    ) => {
      state.items = action.payload;
    },
    setPages: (
      state: CharacterItemsStateType,
      action: PayloadAction<number>
    ) => {
      state.pages = action.payload;
    },
  },
});

export const { setCharacterItems, setPages } = characterItemsSlice.actions;

export const getCharacterItemsState = (state: {
  characterItems: CharacterItemsStateType;
}) => state.characterItems;

export const getCharacterItems = createSelector(
  [getCharacterItemsState],
  (characterItemsState) => characterItemsState.items
);
export const getCharacterPages = createSelector(
  [getCharacterItemsState],
  (characterItemsState) => characterItemsState.pages
);

export default characterItemsSlice.reducer;
