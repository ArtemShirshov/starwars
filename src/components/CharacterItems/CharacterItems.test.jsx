import React from 'react';
import renderer from 'react-test-renderer';

import { CharacterItems } from './CharacterItems';

let mockData = {
  loading: false,
  characterItems: [
    {
      name: 'Luke',
      id: '1',
    },
  ],
  handleChangePage: jest.fn(),
  page: 1,
  onClick: jest.fn(),
  characterPages: 8,
};

jest.mock('./CharacterItems.viewmodel', () => ({
  CharacterItemsViewModel: () => mockData,
}));

jest.mock('./CharacterItem/CharacterItem', () => ({
  CharacterItem: 'CharacterItem',
}));
jest.mock('./CharacterItem/CharacterItemSkeleton', () => ({
  CharacterItemSkeleton: 'CharacterItemSkeleton',
}));
jest.mock('@mui/material', () => ({
  ...jest.requireActual('@mui/material'),
  Pagination: 'Pagination',
  Skeleton: 'Skeleton',
}));

describe('CharacterItems', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<CharacterItems />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with loading', () => {
    mockData = {
      ...mockData,
      loading: true,
    };

    const tree = renderer.create(<CharacterItems />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
