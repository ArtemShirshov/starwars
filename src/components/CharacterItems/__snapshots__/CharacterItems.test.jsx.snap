// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterItems renders correctly 1`] = `
Array [
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-iv7vbz-MuiGrid-root"
      onClick={[Function]}
    >
      <CharacterItem
        id="1"
        name="Luke"
      />
    </div>
  </div>,
  <div
    className="MuiBox-root css-b4acux"
  >
    <Pagination
      color="primary"
      count={8}
      onChange={[MockFunction]}
      page={1}
      shape="rounded"
    />
  </div>,
]
`;

exports[`CharacterItems renders correctly with loading 1`] = `
Array [
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-iv7vbz-MuiGrid-root"
    >
      <CharacterItemSkeleton />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-iv7vbz-MuiGrid-root"
    >
      <CharacterItemSkeleton />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-iv7vbz-MuiGrid-root"
    >
      <CharacterItemSkeleton />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-iv7vbz-MuiGrid-root"
    >
      <CharacterItemSkeleton />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-iv7vbz-MuiGrid-root"
    >
      <CharacterItemSkeleton />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-iv7vbz-MuiGrid-root"
    >
      <CharacterItemSkeleton />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-iv7vbz-MuiGrid-root"
    >
      <CharacterItemSkeleton />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-iv7vbz-MuiGrid-root"
    >
      <CharacterItemSkeleton />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-iv7vbz-MuiGrid-root"
    >
      <CharacterItemSkeleton />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3 css-iv7vbz-MuiGrid-root"
    >
      <CharacterItemSkeleton />
    </div>
  </div>,
  <div
    className="MuiBox-root css-b4acux"
  >
    <Pagination
      color="primary"
      count={8}
      onChange={[MockFunction]}
      page={1}
      shape="rounded"
    />
  </div>,
]
`;
